### PR TITLE
Add dependencies and update code

### PR DESCRIPTION
**Pull request by andyherfer**

### CODE REVIEW

**Comment by andresdiverGPT** [comment]
Thank you for providing the changes, Andy. Here are some code review comments and suggestions.

### CommentAgent.py

1. It seems like the `GPTWrapper` class and related code have been removed and replaced with imports from `GPTutils`. Ensure that the functionality you depend upon in `CommentAgent` is still intact with these changes. Also, verify that `GPTutils` has all necessary components after the refactor.

2. Make sure that the functionalities provided by the removed lines, especially the error handling and messaging system, are catered for in the updated classes you are now using.

### GPTutils.py

1. Add docstrings for all new functions and classes, just as you have for existing ones to maintain consistency and provide proper documentation.

```suggestion 
+ Line 110 Side RIGHT:     def __str__(self) -> str:
+ Line 111 Side RIGHT:         """
+ Line 112 Side RIGHT:         Returns a readable string representation of the conversation.
+ Line 113 Side RIGHT: 
+ Line 114 Side RIGHT:         Returns:
+ Line 115 Side RIGHT:             str: A string representation of the conversation.
+ Line 116 Side RIGHT:         """
```

2. Utilize logging instead of plain `print` statements for better production-quality code and flexibility in the output.

```suggestion 
+ Line 140 Side RIGHT:         print("splitted code docs")
+ Line 151 Side RIGHT:         print("splitted text docs")
```

Replace `print` with logging:

```suggestion 
+ Line 140 Side RIGHT:         logging.info("Splitted code docs")
+ Line 151 Side RIGHT:         logging.info("Splitted text docs")
```

3. Ensure proper testing of the `cosim` and `cosim_matrix` functions as these are critical for the semantic search functionality in `FilesIndex`.

### GithubHandlers.py

1. The additional check for `patch is None` seems like a safeguard. However, ensure that this code path is tested in case of handling PRs with no file changes, to return a meaningful response instead of an empty delta block.

```suggestion 
+ Line 227 Side RIGHT:         if patch is None:
+ Line 228 Side RIGHT:             return f"No Changes in {filename}\n" + "-" * 30 + "\n"
```

In general, make sure to run integration tests to ensure the new modules being imported work well with existing code. This should include checking for any interface or signature changes that can affect the current functionality and handling backward compatibility if needed.

**Comment by andresdiverGPT** [comment]
### CommentAgent.py

1. It's good to see dead code being cleaned up, however, ensure that any useful functionality or error handling previously in the `GPTWrapper` class has been properly migrated to the new `GPTutils` and `GithubHandlers` modules.

```suggestion
Line 5 Side RIGHT: # Ensure that the new imports provide equivalent functionality
from GPTutils import GPTWrapper
from GithubHandlers import GithubHandler
```

2. The removal of direct `openai` module usage seems intentional for better modularity. Just double-check that the `GPTWrapper` class in `GPTutils` maintains all the necessary logic required by `CommentAgent`.

### GPTutils.py

1. It looks like the `GPTWrapper` implementation was moved to this file from `CommentAgent.py`. Moving related functionality to a utilities file is a good practice for maintainability and code organization.

2. Ensure that all methods within `GPTWrapper` are used and relevant to your application; otherwise, consider removing unused methods to keep the codebase lean.

3. Remember to add exception handling to API calls, such as the one in `get_response`, to gracefully manage potential failures in the OpenAI service.

```suggestion
Line 81 Side RIGHT:         try:
Line 82 Side RIGHT:             response = self.client.chat.completions.create(
Line 83 Side RIGHT:                 model=self.model, messages=self.conversation, max_tokens=max_tokens
Line 84 Side RIGHT:             )
Line 85 Side RIGHT:             return response.choices[0].message.content
Line 86 Side RIGHT:         except openai.Error as e:
Line 87 Side RIGHT:             logging.error(f"OpenAI API call failed: {e}")
Line 88 Side RIGHT:             return None
```

### GithubHandlers.py

1. The `index_pr` method is a new addition to pre-compute and cache file contents. Ensure this doesn't negatively impact performance for large PRs due to upfront loading of all files.

2. Verify that file content indexing with `FilesIndex` works correctly with the types of files that your project includes.

3. The `get_file_delta` now handles the case where `patch` is None, which can happen if the file was not modified. Ensure you test these cases and provide useful feedback to users.

```suggestion
Line 264 Side RIGHT:         if patch is None:
Line 265 Side RIGHT:             return f"{filename} - No changes detected\n" + "-" * 30 + "\n"
```

4. The `get_file_context` seems to extract context from the repository's existing code to better understand the changes. This is a clever approach but verify the performance implications and correct extraction of relevant context.

Overall, ensure comprehensive testing of the changes, including unit tests for newly added methods and classes, and integration tests to verify that the updated code interacts properly with other systems it depends on. Additionally, perform a review of the logging and exception handling to make certain that potential issues can be diagnosed and debuged effectively.

File: CommentAgent.py [hunk 1]
```diff
@@ -1,119 +1,13 @@
...
-        Returns a readable string representation of the conversation.
-
-        Returns:
-            str: A string representation of the conversation.
-        """
-        return "\n".join(
-            f"{msg['role']}: {msg['content']}" for msg in self.conversation
-        )
+from GithubHandlers import GithubHandler
+from GPTutils import GPTWrapper
```
Comment (andresdiverGPT) [inline CommentAgent.py:10]: Ensure that `GPTWrapper` from `GPTutils` and `GithubHandler` are providing all the necessary functionalities as before refactoring. Missing any former implementation might cause regression in the workflow of `CommentAgent` class.
```suggestion
        from GPTutils import GPTWrapper
        from GithubHandlers import GithubHandler
```

File: GPTutils.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        Sets a message in the conversation by index.
+
+        Args:
+            index (int): The index of the message to set.
+            value (tuple[str, str]): A tuple containing the role and content of the new message.
+        """
+        role, content = value
+        self.edit_message(index, role, content)
+
+    def __str__(self) -> str:
```
Comment (andresdiverGPT) [inline GPTutils.py:110]: Please add docstrings for all new functions and classes, similar to the provided example, to maintain consistency and thorough documentation throughout the project.
```suggestion
        def __str__(self) -> str:
            """
            Returns a readable string representation of the conversation.

            Returns:
                str: A string representation of the conversation.
            """
```
        Please apply this suggestion across all new methods and classes introduced in this PR.

File: GPTutils.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        """
+        Gets a response from the API based on the current conversation.
+
+        Args:
+            max_tokens (int, optional): The maximum number of tokens to generate in the response. Defaults to 1000.
+
+        Returns:
+            str: The generated response.
+        """
+        response = self.client.chat.completions.create(
```
Comment (andresdiverGPT) [inline GPTutils.py:81]: It is essential to handle potential API failures gracefully. Please surround the API call with try-except block to catch and log any exceptions that might occur during the call to OpenAI.
```suggestion
        try:
            response = self.client.chat.completions.create(
                model=self.model, messages=self.conversation, max_tokens=max_tokens
            )
            return response.choices[0].message.content
        except openai.Error as e:
            logging.error(f"OpenAI API call failed: {e}")
            return None
```

File: GithubHandlers.py [hunk 4]
```diff
@@ -152,29 +164,54 @@ def get_file_contents(self, filepath: str, ref=None):
             traceback.print_exc()
             return "Could not retrieve file contents."
 
-    def get_pr_deltas(self, pr):
+    def get_pr_deltas(self, pr) -> str:
```
Comment (andresdiverGPT) [inline GithubHandlers.py:167]: Refactor the `get_pr_deltas` method to ensure useful feedback is provided to the users in scenarios where a file may not have been modified. Consider also handling large PRs efficiently to prevent performance issues.
```suggestion
        def get_pr_deltas(self, pr) -> str:
            """
            Fetches details about the changes in a Pull Request and provides additional context for better understanding.
            
            Args:
                pr (PullRequest): The pull request to get deltas for.
            
            Returns:
                str: A comprehensive overview of the PR details.
            """
            # Code implementation continues...
```
        Make sure to test these edge cases thoroughly.

**Review by andresdiverGPT — CHANGES_REQUESTED** [review]
The PR introduces several critical changes and refactors. It's paramount to ensure functionality equivalence after module re-organization. Please make sure that all functionalities previously contained in the GPTWrapper class, especially related to error handling and API interaction, are preserved in the new GPTutils and GithubHandlers modules. Additionally, thorough testing should be conducted to ensure integration with other systems is not adversely affected.

File: CommentAgent.py (hunk 1 above)
Comment (andresdiverGPT) [inline CommentAgent.py:10]: Ensure that the classes `GPTWrapper` and `GithubHandler` from the imported modules provide all of the functionalities used in this class.

File: GPTutils.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+        Args:
+            model (str, optional): The name of the GPT model to use. Defaults to "text-davinci-002".
+        """
+        self.client = OpenAI()
+        self.model: str = model
+        self.conversation: list[
+            dict[str, str]
+        ] = []  # This will hold our conversation messages
+
+    def add_message(self, role: str, content: str) -> None:
```
Comment (andresdiverGPT) [inline GPTutils.py:44]: Please add documentation for the `add_message` method.

File: GPTutils.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+        """
+        Adds a message to the conversation history.
+
+        Args:
+            role (str): Indicates who is acting on the conversation [system, user, assistant].
+            content (str): The actual message.
+        """
+        self.conversation.append({"role": role, "content": content})
+
+    def edit_message(self, index: int, role: str, content: str) -> None:
```
Comment (andresdiverGPT) [inline GPTutils.py:54]: Utilize proper logging instead of print statements for debug output.

File: GPTutils.py [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+        Raises:
+            IndexError: If the index is out of range.
+        """
+        if index < len(self.conversation):
+            self.conversation[index] = {"role": role, "content": content}
+        else:
+            raise IndexError("Message index out of range")
+
+    def get_response(self, max_tokens: int = 1000) -> str:
```
Comment (andresdiverGPT) [inline GPTutils.py:71]: Make sure to handle potential exceptions when calling external APIs in `get_response`.

File: GithubHandlers.py [hunk 8]
```diff
@@ -49,6 +50,17 @@ def __init__(
         self.prs_nums = [pr.number for pr in self.prs]
         self.prs_dict = {k: v for k, v in zip(self.prs_nums, self.prs)}
 
+    def index_pr(self, pr):
```
Comment (andresdiverGPT) [inline GithubHandlers.py:53]: Verify if the `index_pr` method could impact performance for large PRs and consider optimization.

File: GithubHandlers.py (hunk 4 above)
Comment (andresdiverGPT) [inline GithubHandlers.py:167]: When a patch is `None`, ensure to provide a meaningful message to the end user, and test these scenarios.

**Review by andresdiverGPT — CHANGES_REQUESTED** [review]
Please review the inline suggestions to ensure code quality and consistency.

File: GPTutils.py [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+    "java": "java",
+    "html": "html",
+    "md": "markdown",
+}
+
+
+class GPTWrapper:
+    def __init__(self, model: str = GPTsettings.MODEL) -> None:
+        """
+        Initializes a new instance of the GPTWrapper class.
```
Comment (andresdiverGPT) [inline GPTutils.py:33]: Missing docstrings for new functions and classes, please add them to maintain proper documentation.

File: GPTutils.py [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+        self.index_files()
+
+    def index_files(self) -> None:
+        """
+        Indexes the files in the `files` dictionary by creating a mapping between each file name and a list of document
+        indices in the `docs` list. Also computes embeddings for each document and stores them in the `embeddings` array.
+        """
+        embeddings = []
+        last_index = 0
+        for filename, content in self.files.items():
```
Comment (andresdiverGPT) [inline GPTutils.py:211]: Consider adding logging for better production-quality code and flexibility in the output.

File: GithubHandlers.py [hunk 11]
```diff
@@ -152,29 +164,54 @@ def get_file_contents(self, filepath: str, ref=None):
...
+    def get_pr_deltas(self, pr) -> str:
         """
         Gets the deltas for a pull request.
 
         Args:
             pr (PullRequest): The pull request to get deltas for.
 
         Returns:
-            str: The deltas for the pull request.
+            str: The deltas for the pull request plus additional code context.
```
Comment (andresdiverGPT) [inline GithubHandlers.py:175]: Ensure that the new path for `patch is None` in `get_file_delta` is tested in case of PRs with no file changes, returning a meaningful response.

**Review by andresdiverGPT — CHANGES_REQUESTED** [review]
Thank you for providing the changes, Andy. 

Please ensure that:
- The functionalities provided by the removed lines, especially error handling and messaging system, are catered for in the updated classes you are now using.
- All methods within `GPTWrapper` are used and relevant to your application; otherwise, consider removing unused methods to keep the codebase lean.
- Exception handling is added to API calls in `GPTWrapper.get_response`.
- The `index_pr` method performance has been tested with large PRs since it pre-fetches all file contents.
- The context extraction for file deltas is properly tested and verified for performance and accuracy.

Remember to add docstrings to all new functions and classes, and to properly test your changes, including the critical `cosim` and `cosim_matrix` functions for semantic search functionality.

File: GPTutils.py [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+            {Filename: FileContent}
+        """
+        self.files = files
+        self.index: Dict[str, List[int]] = {}
+        self.idx_to_filename = {}
+        self.embeddings_model = OpenAIEmbeddings()
+        self.docs: List[str] = []
+        self.embeddings: np.ndarray = None
+        self.index_files()
+
```
Comment (andresdiverGPT) [inline GPTutils.py:203]: Please make sure to add appropriate error handling to newly introduced functions in case of unexpected behaviors, for instance in asynchronous calls.
```suggestion
    def cosim(a, b):
        try:
            return np.dot(a, b) / (norm(a) * norm(b))
        except Exception as e:
            logging.error(f"Failed to calculate cosine similarity: {str(e)}")
            return 0.0
```

File: GPTutils.py [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+        elif isinstance(queries, str):
+            query = queries  # signaling that queries is a string
+            query_embedding = np.array(self.embeddings_model.embed_query(query))
+            scores = cosim_matrix(query_embedding, self.embeddings)
+
+        top_k_idxs = np.argsort(scores)[::-1][:top_k]
+
+        if sorted_by == "file":
+            top_k_idxs = sorted(top_k_idxs)
+
```
Comment (andresdiverGPT) [inline GPTutils.py:258]: Incorporate exception catching for OpenAI client interactions to avoid unexpected crashes and to handle API limitations or downtime gracefully, for example:
```suggestion
    def get_response(self, max_tokens: int = 1000) -> str:
        try:
            response = self.client.chat.completions.create(
                model=self.model, messages=self.conversation, max_tokens=max_tokens
            )
            return response.choices[0].message.content
        except openai.OpenAIError as e:
            logging.error(f"OpenAI API error: {e}")
            return ""
```

File: CommentAgent.py [hunk 14]
```diff
@@ -360,12 +254,15 @@ def review_pr(self) -> bool:
         for _ in range(3):
             message_format = GPTsettings.MESSAGE_FORMAT
             self.GPT.add_message("user", message_format)
-            message_response = self.GPT.get_response(max_tokens=2500)
+            message_response = self.GPT.get_response(
```
Comment (andresdiverGPT) [inline CommentAgent.py:257]: The change in the `max_tokens` default value is quite large. It's fine as long as you've checked the implications this will have on the behavior of your bot, and you're within the API's limits.

File: CommentAgent.py [hunk 15]
```diff
@@ -360,12 +254,15 @@ def review_pr(self) -> bool:
...
+                max_tokens=4096
+            )  # Max length of response is 4096 (max allowed for responses by openai)
             try:
                 logging.info(message_response)
                 response_parser = ResponseParser(message_response)
                 body, event, comments = response_parser.get_body_event_and_comments()
                 self.pr.create_review(body=body, event=event, comments=comments)
+                print(self.GPT)
```
Comment (andresdiverGPT) [inline CommentAgent.py:265]: This print statement could be replaced with a logging call for better control of the output, consistency, and logging levels.
```suggestion
    logging.info(str(self.GPT))
```

File: GithubHandlers.py [hunk 16]
```diff
@@ -224,6 +261,8 @@ def get_file_delta(self, patch, filename):
         Returns:
             str: The deltas for the file formatted for position reviews.
         """
+        if patch is None:
+            return f"{filename}\n" + "-" * 30 + "\n"
```
Comment (andresdiverGPT) [inline GithubHandlers.py:265]: Consider handling the case where `patch` is `None` more gracefully. For example, you could indicate that no changes were detected between the base and head branches for the given file.
```suggestion
    if patch is None:
        return f"{filename} - No changes detected between the base and head branches.
" + "-" * 30 + "
"
```

File: GithubHandlers.py [hunk 17]
```diff
@@ -152,29 +164,54 @@ def get_file_contents(self, filepath: str, ref=None):
...
+        return delta_query
+
+    def get_delta_query_block(self, file_delta_str, file_delta_context):
+        query_block = f"""Add a code review comment to the following changes:
+        {file_delta_str}
+        
+        Only for context, here is some code from the repository (DO NOT REVIEW THIS CODE, IT IS ONLY FOR CONTEXTUAL INPUT):
+        {file_delta_context}\n"""
+        return query_block
+
```
Comment (andresdiverGPT) [inline GithubHandlers.py:200]: The method `get_file_context` is adding context to the code reviews, but ensure that this additional context is optimized for both relevance and performance.

**Review by andresdiverGPT — CHANGES_REQUESTED** [review]
Hi Andy, you made significant changes in this PR. Here are some suggestions and points to note:
